Add reserved snapshot/repo action

grcevski opened the pull request:
This PR adds support for /_snapshot/repo file based settings.

Pre-requisite for: https://github.com/elastic/elasticsearch/pull/89567

Relates to https://github.com/elastic/elasticsearch/issues/89183

Commit:
Add reserved snapshot/repo action

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

elasticsearchmachine commented:
Hi @grcevski, I've created a changelog YAML for you.

grcevski committed:
Update docs/changelog/89601.yaml

grcevski commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/delete/TransportDeleteRepositoryAction.java:
```diff
@@ -69,4 +73,14 @@ protected void masterOperation(
         );
     }
+
+    @Override
```
SLM and ILM might depend on this snapshot repo configuration, but not always. This extends the handler interface to allow optional dependencies, they are here simply for ordering purposes so we can create the repo before the SLM policy.

grcevski commented on server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java:
```diff
@@ -193,52 +184,7 @@ public void registerRepository(final PutRepositoryRequest request, final ActionL
-                return ClusterState.builder(currentState).metadata(mdBuilder).build();
-            }
-
+        submitUnbatchedTask("put_repository [" + request.name() + "]", new RegisterRepositoryTask(this, request, acknowledgementStep) {
```
This refactoring simply moves the code so we can reuse the task execute method for the reserved state handler.

grcevski commented on server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java:
```diff
@@ -142,16 +142,7 @@ public RepositoriesService(
-            listener.onFailure(e);
-            return;
-        }
+        validateRepository(request, listener);
```
The validation logic was extracted so we can use it in the reserved state handler.

original-brownbear replied:
This seems broken, we should return here on any validation exception?
Maybe better to leave this a non-async method and do the try catch here and let the exception bubble up in the new spot this is used in where the listener is somewhat artificial anyway?

grcevski replied:
So while fixing this I noticed something, the name validation is a sync failure, but creating the repository uses the listener. I'm not sure there's a reason for this difference, but I changed the code to mimic the original behaviour now.

grcevski commented on server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java:
```diff
@@ -350,40 +390,7 @@ public void clusterStateProcessed(ClusterState oldState, ClusterState newState)
-                throw new RepositoryMissingException(request.name());
-            }
-
+        submitUnbatchedTask("delete_repository [" + request.name() + "]", new UnregisterRepositoryTask(request, listener) {
```
Similar extraction of the execute logic in a standalone task to be reused.

grcevski commented on server/src/main/java/org/elasticsearch/reservedstate/ReservedClusterStateHandler.java:
```diff
@@ -77,6 +77,20 @@ default Collection<String> dependencies() {
+     *
+     * @return a collection of optional reserved state handler names
+     */
+    default Collection<String> optionalDependencies() {
```
SLM and ILM might depend on this snapshot repo configuration, but not always. This extends the handler interface to allow optional dependencies, they are here simply for ordering purposes so we can create the repo before the SLM policy.

grcevski commented on server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:
```diff
@@ -82,13 +82,11 @@ public FileSettingsService(ClusterService clusterService, ReservedClusterStateSe
 
-    // package private for testing
-    Path operatorSettingsDir() {
+    public Path operatorSettingsDir() {
```
Making these public so that we can use them in tests outside of the package.

grcevski commented on server/src/main/java/org/elasticsearch/reservedstate/service/ReservedClusterStateService.java:
```diff
@@ -267,7 +267,21 @@ private void addStateHandler(String key, Set<String> keys, LinkedHashSet<String>
         ordered.add(key);
     }
+
+    /**
```
The state handlers are initialized on creation of the service, which is hung off the ActionModule. The Node creation code has many service interdependencies and it's impossible to create all reserved state handlers ahead of time. This API allows us the flexibility to add other state handlers as we build the modules in Node.

original-brownbear reviewed (commented):
Gave it a quick read and looks reasonable. Will do a deeper review soon, but I think found one issue that needs addressing until then.

Commit:
Fix bug on validate refactor

Commit:
Merge branch 'operator/repo' of github.com:grcevski/elasticsearch into operator/repo

Commit:
Extract the name validation early

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/reservedstate/ReservedRepositoryAction.java:
```diff
@@ -0,0 +1,112 @@
+
+        Map<String, ?> source = parser.map();
+
+        for (String name : source.keySet()) {
```
Not too important but this is a bit of a strange loop, why not loop EntrySet and avoid the `source.get(name)`?

original-brownbear reviewed (commented):
Thanks Nikola, this looks alright to me now in terms of the cluster state updates and changes to the `RepositoriesService`. 
But just generally speaking, shouldn't this have some end-to-end test in the form of an internal cluster test or REST test?

grcevski commented:
> Thanks Nikola, this looks alright to me now in terms of the cluster state updates and changes to the `RepositoriesService`.
> But just generally speaking, shouldn't this have some end-to-end test in the form of an internal cluster test or REST test?

Thanks Armin! Great point on the testing, I had a test when this work was together with the SLM policies, I forgot to bring the part of the integration test. I'll follow-up with and update now.

Commit:
Merge branch 'main' into operator/repo

Commit:
Address PR comment.

Commit:
Add integration test

Commit:
Merge branch 'main' into operator/repo

Commit:
Fix deadlock bug exposed by the test

The new test exposed a very rare bug where the
file settings service was in the middle of processing
the file when the node closed. This terminated the
cluster state update task, but nobody unlocked the
latch await. The fix allows the stop operation to
properly terminate the watcher thread.

Test added that exposes the bug.

grcevski commented:
There's still a race condition. I need to think a bit more how to prevent it.

Commit:
Fix race condition exposed by the test

There was one more race condition related to service
stop. The setting of the processing latch raced against
the stop method execution. The stop method may have
still seen the processing latch as null or another stale
instance. With this fix after we receive a processing
latch, we check if the watcher state is still valid, and
only then wait.

Another test added consistently exposing the timing hole.

Commit:
Merge branch 'main' into operator/repo

original-brownbear commented on server/src/internalClusterTest/java/org/elasticsearch/reservedstate/service/RepositoriesFileSettingsIT.java:
```diff
@@ -0,0 +1,250 @@
+        );
+
+        // This should succeed, nothing was reserved
+        client().execute(PutRepositoryAction.INSTANCE, sampleRestRequest("err-repo")).actionGet();
```
Can we use just plain `get()`, `actionGet()` hides the exact point where the exception was thrown by unwrapping the execution exception and unless it's needed for something like `expectThrows` above I'd rather avoid it because it makes debugging test failures a pain :)

original-brownbear commented on server/src/internalClusterTest/java/org/elasticsearch/reservedstate/service/RepositoriesFileSettingsIT.java:
```diff
@@ -0,0 +1,250 @@
+        final var reposResponse = client().execute(
+            GetRepositoriesAction.INSTANCE,
+            new GetRepositoriesRequest(new String[] { "repo", "repo1" })
+        ).actionGet();
```
Lets use `get` here

original-brownbear commented on server/src/internalClusterTest/java/org/elasticsearch/reservedstate/service/RepositoriesFileSettingsIT.java:
```diff
@@ -0,0 +1,250 @@
+
+    private void assertMasterNode(Client client, String node) {
+        assertThat(
+            client.admin().cluster().prepareState().execute().actionGet().getState().nodes().getMasterNode().getName(),
```
`get` :)

original-brownbear reviewed (approved):
LGTM just one thing to maybe fix in the tests if you have a sec :) Thanks for the iterations!

Commit:
Switch to using get().

Commit:
spotless

grcevski commented:
Thanks Armin!
